components/brave_rewards/resources/page/reducers/promotion_reducer.ts

```diff
@@ -0,0 +1,234 @@
+        if (promotionId === item.promotionId) {
+          item.captchaImage = captchaImage
+          item.captchaId = captchaId
+          item.hint = hint
+        }
+        return item
+      })
+
+      state = {
+        ...state,
+        promotions
+      }
+      break
+    }
+    case types.ATTEST_PROMOTION: {
+      const promotionId = payload.promotionId
+
+      if (!promotionId || !payload.x || !payload.y) {
+        break
+      }
+
+      const currentPromotion = getPromotion(promotionId, state.promotions)
+
+      if (!currentPromotion || !currentPromotion.captchaId) {
+        break
+      }
+
+      chrome.send('brave_rewards.attestPromotion', [promotionId, JSON.stringify({
+        captchaId: currentPromotion.captchaId,
+        x: parseInt(payload.x, 10),
+        y: parseInt(payload.y, 10)
+      })])
+      break
+    }
+    case types.RESET_PROMOTION: {
+      const promotionId = payload.promotionId
+      if (!state.promotions || !promotionId) {
+        break
+      }
+
+      const currentPromotion = getPromotion(promotionId, state.promotions)
+
+      if (!currentPromotion) {
+        break
+      }
+
+      const promotions = state.promotions.map((item: Rewards.Promotion) => {
+        if (currentPromotion.promotionId === item.promotionId) {
+          item.captchaStatus = null
+          item.captchaImage = ''
+          item.captchaId = ''
+          item.hint = ''
+        }
+        return item
+      })
+
+      state = {
+        ...state,
+        promotions
+      }
+      break
+    }
+    case types.DELETE_PROMOTION: {
+      const promotionId = payload.promotionId
+      if (!state.promotions || !promotionId) {
+        break
+      }
+      const currentPromotion = getPromotion(promotionId, state.promotions)
+
+      if (!currentPromotion) {
+        break
+      }
+
+      let promotions = state.promotions.map((item: Rewards.Promotion) => {
+        if (currentPromotion.promotionId === item.promotionId) {
+          item.captchaStatus = null
+          item.status = 4
+        }
+
+        return item
+      })
+
+      state = {
+        ...state,
+        promotions
+      }
+      break
+    }
+    case types.ON_PROMOTION_FINISH: {
+      state = { ...state }
+      let newPromotion: any = {}
+      const result = payload.properties.result
+
+      if (!state.promotions) {
+        break
+      }
+
+      let promotionId = payload.properties.promotion.promotionId
+      newPromotion.promotionId = promotionId
+
+      switch (result) {
+        case 0:
+          let ui = state.ui
+          newPromotion.captchaStatus = 'finished'
+          newPromotion.captchaImage = ''
+          newPromotion.captchaId = ''
+          newPromotion.hint = ''
+          ui.emptyWallet = false
+
+          state = {
+            ...state,
+            ui
+          }
+
+          chrome.send('brave_rewards.getWalletProperties')
+          chrome.send('brave_rewards.fetchBalance')
+          break
+        case 6:
```

Comment:
Const's for these notable positions?

Reply:
right now we can't use direct definition